### PR TITLE
fix: Resolve and validate registered custom classes used as types in model fields.

### PR DESCRIPTION
**Pull request by Isakdl**
# Changes

Fixes regression that gave error on valid custom classes, this applies to all generic types too.

Example:

```yaml
# generator.yaml
...
extraClasses:
  - package:serverpod_test_shared/serverpod_test_shared.dart:ExternalCustomClass
```

```yaml
# models/example.yaml
class: Example
fields:
  customClassField: ExternalCustomClass
  list: List<ExternalCustomClass>
  map: Map<String, ExternalCustomClass>
```

All of the above declarations gave an invalid type error when running `serverpod generate`.

With this PR all of the above will not give any error and will resolve the import paths correctly.

Closes: https://github.com/serverpod/serverpod/issues/1932

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes
None

### CODE REVIEW

**Comment by Isakdl** [comment]
@AndryHTC PR with a fix for your issue here. You should not have to change any model files after this change.

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/field_datatype_test.dart [hunk 1]
```diff
@@ -1032,6 +1033,176 @@ void main() {
...
+      var definition = definitions.first as ClassDefinition;
+      expect(
+        definition.fields.first.type.url,
+        'package:shared_package/src/lib/custom_example.dart',
+      );
+    });
+  });
+
+  group(
+      'Given a class with a type set to a list of the class name of a custom type',
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/field_datatype_test.dart:1087]: Suggestion(Non blocking): Change to: `Given a class with a type set to a list of custom classes`

Comment (SandPod) [reply tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/field/field_datatype_test.dart:1087]: Suggestion for the same name change for the other map test.

**Comment by AndryHTC** [comment]
Awesome job. Any ETA for the release?

File: tools/serverpod_cli/lib/src/analyzer/models/model_analyzer.dart [hunk 2]
```diff
@@ -38,6 +38,7 @@ class SerializableModelAnalyzer {
   /// Best effort attempt to extract an model definition from a yaml file.
   static SerializableModelDefinition? extractModelDefinition(
     ModelSource modelSource,
+    GeneratorConfig config,
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/models/model_analyzer.dart:41]: Suggestion: only take the extraClasses as parameter.

If the extra classes is the only field that we need from the config I suggest we don't create a dependency towards the config here but instead make clear what this method needs by defining it as a parameter of the method.

**Review by SandPod — APPROVED** [review]
Nice work!

One suggestion I think should be resolved, the other one is optional.
(Best to make it as clear as possible what we actually read from the config).

**Pull request merged**